### DISCUSSION TITLE
[CodeStyle] Use Unity events [...]

### DISCUSSION THREAD

**Discussion by ViMaSter** [Studio-Lovelies/GG-JointJustice-Unity#167] [Ideas]
> Use Unity events to notify other parts of the code so we can hook things together in a more modular way

---

I think this is merely giving it formal recognition, but I think we decided to move away from events and make heavy use of `interfaces`, and non-`MonoBehaviour` derivatives instead of events and make heavy use of automated testing.

How about the following update?

> - When updating existing functionality, preferably de-couple custom logic from Unity-specific functionality unless strictly required
> - When adding new functionality, adding a test scene to be able to play with the new feature with one click is highly encouraged
> - It is encouraged to create automated tests alongside any changes or newly created functionality
>   - Covering the [happy path](https://en.wikipedia.org/wiki/Happy_path) is a good start
>   - If you encounter realistic edge-cases while working, consider also adding automatic tests for those
>   - Prefer writing EditMode tests over PlayMode tests
>   - [Use: `yield return null` rather than `yield return new WaitForEndOfFrame()`](http://answers.unity.com/answers/755208/view.html)
>   - Use `yield return TestTools.WaitForState(...)` rather than `yield return new WaitForSeconds(...)`

**Comment by PhoebeMitchell** [comment]
Update looks good. I fixed some grammar. Corrected a missing space before 'yield', and 'automatically tests' to 'automatic tests'.

If an event is required what event should be preferred? C# or Unity? Previously we said always use UnityEvents. As C# events are used in ActionDecoder maybe change it to say only use UnityEvents if they need to be exposed in the inspector?

**Reply by ViMaSter** [reply]
> only use UnityEvents if they need to be exposed in the inspector

...sounds like a good rule of thumb to me!  

Generally I haven't used any in the code I've contributed so far, and I remember @notexplosive having an opinion on it.  
@notexplosive Any idea how to make that into a rule we can add here?
